### PR TITLE
Add support for signing attestation with Fulcio

### PR DESCRIPTION
**Pull request by colek42**

### CODE REVIEW

File: cmd/witness/cmd/keyloader.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	"github.com/testifysec/witness/pkg/signer/file"
+	"github.com/testifysec/witness/pkg/signer/fulcio"
+	"github.com/testifysec/witness/pkg/signer/spiffe"
+)
+
+func getSigners(ko options.KeyOptions) ([]cryptoutil.Signer, []error) {
+	signers := []cryptoutil.Signer{}
+	errors := []error{}
+
+	ctx := context.Background()
```
Comment (mikhailswift) [inline cmd/witness/cmd/keyloader.go:32]: is there a reason to shuffle this to inlined vs a parameter and renaming this function?  not really a big deal.  my idea behind having this context be a parameter was the ability to add timeouts/etc at the call site.

was trying to use the language `loadXxx` for functions that opened files/etc.

guess it doesn't matter much.

File: cmd/witness/cmd/verify_test.go [hunk 2]
```diff
@@ -17,17 +17,109 @@ package cmd
...
+	caBytes, err := ioutil.ReadFile(ca.Name())
+	require.NoError(t, err)
+
+	policy := makepolicyCA(t, caBytes)
+	signedPolicy, pub := signPolicyRSA(t, policy)
+
+	workingDir := t.TempDir()
+	attestationDir := t.TempDir()
+
+	err = os.WriteFile(workingDir+"signed-policy.json", signedPolicy, 0644)
```
Comment (mikhailswift) [inline cmd/witness/cmd/verify_test.go:54]: make a habit of using filepath.Join for these operations. less error prone and will handle directory separators or lack of them for you.

File: pkg/cryptoutil/verifier.go [hunk 3]
```diff
@@ -97,3 +99,33 @@ func NewVerifierFromReader(r io.Reader, opts ...VerifierOption) (Verifier, error
 
 	return NewVerifier(key, opts...)
 }
+
+func NewCAVerifier(certs []*x509.Certificate, opts ...VerifierOption) Verifier {
```
Comment (mikhailswift) [inline pkg/cryptoutil/verifier.go:103]: This should be unnecessary with the existing X509Verifier, also I see no usages of this

File: cmd/witness/cmd/keyloader.go (hunk 1 above)
Comment (colek42) [reply cmd/witness/cmd/keyloader.go:32]: yeah, those changes don't make sense.  I'll change back to load and add the ctx as a parameter.

File: cmd/witness/cmd/sign.go [hunk 4]
```diff
@@ -47,6 +47,12 @@ func SignCmd() *cobra.Command {
 //we need to abstract where keys are coming from, etc
 func runSign(so options.SignOptions) error {
 	ctx := context.Background()
+
+	if so.KeyOptions.FulcioURL != "" {
```
Comment (mikhailswift) [inline cmd/witness/cmd/sign.go:51]: we shouldn't add the option to sign for fulcio if it's not supported

**Review by mikhailswift — APPROVED** [review]

File: cmd/witness/cmd/sign.go (hunk 4 above)
Comment (colek42) [reply cmd/witness/cmd/sign.go:51]: we can sign, just cant verify yet!

Comment (mikhailswift) [reply cmd/witness/cmd/sign.go:51]: It looks like we’re throwing an error if witness run is run with the fulcio url option, am I misreading it?

Comment (colek42) [reply cmd/witness/cmd/sign.go:51]: Only for sign.  Fulcio options are in the KeyOptions struct.  We can add support for signing with fulcio, but we also need ca verification of policy for this to make sense.

**Pull request merged**